#### Project: CropGuard AI

The project uses two supervised ML tasks.

1. 🌾 Pre-Cultivation Crop Yield Prediction
ML Task: Regression
Target (Y): yield — existing numerical column
Purpose: Predict the expected crop yield in tonnes/hectare for a selected district, crop, season, and agricultural year using historical crop performance, rainfall, irrigation, and contextual information.
Output: Predicted Yield
Example: 4.82 tonnes/hectare
2. ⚠️ Agricultural Risk Classification
ML Task: Classification
Target (Y): risk_level — derived target
Classes: Low Risk / Moderate Risk / High Risk
Purpose: Identify the agricultural risk associated with a particular district–crop–season combination using historical crop performance, rainfall conditions, irrigation conditions, and other validated features.
Output: Risk Level + Risk Probability
Example: HIGH RISK — 78% probability

In [7]:
### coding 

In [151]:
## importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [152]:
import os 
path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
print(os.path.exists(path))

True


In [153]:
df = pd.read_csv(path)
print("Shape:", df.shape)

Shape: (10708, 29)


In [154]:
### 📤 0. Data Loading (from db load tables data as pandas df)


In [155]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus

username = "root"
password = input("Enter MySQL password: ")

engine = create_engine(
    f"mysql+pymysql://{username}:{quote_plus(password)}@localhost:3306/cropguard_ai"
)

with engine.connect() as connection:
    print("Database connection successful!")

Enter MySQL password:  Raju8008


Database connection successful!


In [156]:
import os 
path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"
print(os.path.exists(path))

True


In [157]:
df.to_sql(
    "validated_agriculture_data",
    con=engine,
    if_exists="replace",
    index=False
)

print("Data imported successfully!")

Data imported successfully!


In [158]:
df = pd.read_sql(
    """
    SELECT *
    FROM validated_agriculture_data
    """,
    engine
)

print("Data loaded from MySQL successfully!")
print("Shape:", df.shape)
df.head()

Data loaded from MySQL successfully!
Shape: (10708, 29)


,id,year,state_name,state_code,district_name,district_code,crop_name,crop_code,crop_type,season,...,normal_rainfall,rainfall_deviation,agri_year,irrigation_food_category,irrigation_crop_type,irrigation_crop_name,irrigation_crop_code,irrigation_season,crop_irrigation_area,total_irrigated_area
0,385411,2019-2020,Telangana,36,Adilabad,501,Bajra,103.0,Cereals,Kharif,...,1103.8,69.931180,2019-2020,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79570.0,1048794.0
1,385812,2019-2020,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.8,69.931180,2019-2020,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79570.0,1048794.0
2,407355,2020-2021,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.6,-15.100530,2020-2021,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79635.0,1101556.0
3,450641,2022-2023,Telangana,36,Kumuram Bheem Asifabad,699,Bajra,103.0,Cereals,Kharif,...,1103.6,68.352253,2022-2023,Food Crop,Cereals & Millets,Bajra,103.0,Kharif,79128.0,1040598.0
4,284709,2014-2015,Telangana,36,Adilabad,501,Gram,201.0,Pulses,Rabi,...,1103.6,89.294537,2014-2015,Food Crop,Pulses,Gram,201.0,Rabi,2019.0,271774.0


### CLASSIFICATION TASK — Agrimcultural Risk Classification 

#### 1. Modeling 

 ### 1.1 Pre-Requisites

#### 1.1 Prerequisites – Creating Target Variable (Y)

Before building the classification model, the target variable `risk_level` must be created.

The dataset does not contain a predefined `risk_level` column. Therefore, the target variable is derived from historical agricultural performance, rainfall conditions, and irrigation conditions.

The risk levels are classified into three categories:

- **Low Risk** – agricultural conditions are relatively stable.
- **Moderate Risk** – agricultural conditions show some level of stress.
- **High Risk** – agricultural conditions show significant agricultural stress.

The target variable is created using historical information so that current-year outcome information is not directly used as an input feature. This helps reduce data leakage and makes the risk classification more suitable for prediction.

In [228]:
## Sort Data
df = df.sort_values(
    by=["district_name", "crop_name", "season", "year"]
).reset_index(drop=True)

In [229]:
## Define Group
group_cols = [
    "district_name",
    "crop_name",
    "season"
]

In [230]:
## Previous-year features
df["previous_year_yield"] = (
    df.groupby(group_cols)["yield"].shift(1)
)

df["previous_year_production"] = (
    df.groupby(group_cols)["production"].shift(1)
)

df["previous_year_area"] = (
    df.groupby(group_cols)["area"].shift(1)
)

df["previous_year_irrigation"] = (
    df.groupby(group_cols)["total_irrigated_area"].shift(1)
)

In [231]:
## Historical averages
df["historical_avg_yield"] = (
    df.groupby(group_cols)["yield"]
    .transform(lambda x: x.shift(1).expanding().mean())
)

df["historical_avg_production"] = (
    df.groupby(group_cols)["production"]
    .transform(lambda x: x.shift(1).expanding().mean())
)

df["historical_avg_irrigation"] = (
    df.groupby(group_cols)["total_irrigated_area"]
    .transform(lambda x: x.shift(1).expanding().mean())
)

In [232]:
## Yield trend
df["yield_trend"] = (
    df["previous_year_yield"]
    - df["historical_avg_yield"]
)

In [233]:
## Yield variability
df["yield_variability"] = (
    df.groupby(group_cols)["yield"]
    .transform(
        lambda x: x.shift(1).rolling(
            3, min_periods=2
        ).std()
    )
)

In [234]:
## Area change
df["area_change"] = (
    df["area"] - df["previous_year_area"]
)

In [235]:
## Historical rainfall
df["historical_rainfall"] = (
    df.groupby(group_cols)["actual_rainfall"]
    .transform(
        lambda x: x.shift(1).expanding().mean()
    )
)

In [236]:
## Rainfall anomaly
df["rainfall_anomaly"] = (
    df["actual_rainfall"]
    - df["normal_rainfall"]
)

In [237]:
## Rainfall variability
df["rainfall_variability"] = (
    df.groupby(group_cols)["actual_rainfall"]
    .transform(
        lambda x: x.shift(1).rolling(
            3, min_periods=2
        ).std()
    )
)

In [238]:
## Irrigation change
df["irrigation_change"] = (
    df["total_irrigated_area"]
    - df["previous_year_irrigation"]
)

In [239]:
## Irrigation coverage
df["irrigation_coverage"] = (
    df["total_irrigated_area"] / df["area"]
)

#### Create Risk Level

In [240]:
## Yield stress
df["yield_stress"] = (
    df["historical_avg_yield"]
    - df["previous_year_yield"]
)

In [241]:
## Rainfall stress
df["rainfall_stress"] = (
    -df["rainfall_deviation"]
)

In [242]:
## Irrigation stress
df["irrigation_stress"] = (
    -df["irrigation_change"]
)

In [243]:
## Convert them into percentile-based scores
df["yield_stress_score"] = (
    df["yield_stress"]
    .rank(pct=True)
)

df["rainfall_stress_score"] = (
    df["rainfall_stress"]
    .rank(pct=True)
)

df["irrigation_stress_score"] = (
    df["irrigation_stress"]
    .rank(pct=True)
)

In [244]:
## Combined risk score
df["risk_score"] = (
    0.5 * df["yield_stress_score"]
    + 0.3 * df["rainfall_stress_score"]
    + 0.2 * df["irrigation_stress_score"]
)

#### Create risk level

In [245]:
df["risk_level"] = pd.cut(
    df["risk_score"],
    bins=[-np.inf, 0.33, 0.66, np.inf],
    labels=[
        "Low Risk",
        "Moderate Risk",
        "High Risk"
    ]
)

In [246]:
## Check target
print(df["risk_level"].value_counts())

risk_level
Moderate Risk    4118
Low Risk         1384
High Risk        1186
Name: count, dtype: int64


#### 1.1.1 X & y:
- Picking y output col for Predictive Modelling & considering remaining cols are X

In [247]:
features = [
    "previous_year_yield",
    "historical_avg_yield",
    "yield_trend",
    "yield_variability",
    "previous_year_production",
    "historical_avg_production",
    "previous_year_area",
    "area_change",
    "normal_rainfall",
    "historical_rainfall",
    "rainfall_anomaly",
    "rainfall_deviation",
    "rainfall_variability",
    "previous_year_irrigation",
    "historical_avg_irrigation",
    "irrigation_change",
    "irrigation_coverage",
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]

X = df[features]
y = df["risk_level"]

In [248]:
print("X shape:", X.shape)
print("y shape:", y.shape)


X shape: (8492, 21)
y shape: (8492,)


#### 1.1.2 Train-Test Split:

The dataset is divided into training and testing data based on the `year` column.

Since the agricultural risk classification problem uses historical agricultural data, a **time-based train-test split** is used instead of a random split.

- **Training data:** 2014 to 2021
- **Testing data:** 2022 to 2024

The data from 2014–2021 is used for model learning, while the data from 2022–2024 is kept as unseen future data for testing the trained classification model.

This approach is suitable for the project because it simulates a real-world situation where historical agricultural information is used to predict risk for future years.

In [249]:
## starting year
df["year_start"] = (
    df["agri_year"]
    .astype(str)
    .str.extract(r"(\d{4})")[0]
    .astype(int)
)


In [250]:
df = df.sort_values("year_start").reset_index(drop=True)
years = sorted(df["year_start"].unique())
print("Available agricultural years:")
print(years)

Available agricultural years:
[np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022)]


In [251]:
year_counts = df["year_start"].value_counts().sort_index()
print("Records available for each year:")
display(year_counts)

Records available for each year:


year_start
2015     406
2016     361
2017    1011
2018    1016
2019    1205
2020    1407
2021    1562
2022    1524
Name: count, dtype: int64

In [252]:
## chronological split 
split_index = int(len(years) * 0.80)

train_years = years[:split_index]
test_years = years[split_index:]

print("Training years:")
print(train_years)

print("\nTesting years:")
print(test_years)

Training years:
[np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020)]

Testing years:
[np.int64(2021), np.int64(2022)]


In [253]:
train_mask = df["year_start"].isin(train_years)
test_mask = df["year_start"].isin(test_years)

In [254]:
# Splitting X and y

X_train = X.loc[train_mask].copy()
X_test = X.loc[test_mask].copy()

y_train = y.loc[train_mask].copy()
y_test = y.loc[test_mask].copy()

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (5406, 21)
X_test shape: (3086, 21)
y_train shape: (5406,)
y_test shape: (3086,)


In [255]:
# Check missing values in target

print("Missing values in y_train:", y_train.isna().sum())
print("Missing values in y_test:", y_test.isna().sum())

print("\nTraining target:")
print(y_train.value_counts())

print("\nTesting target:")
print(y_test.value_counts())

Missing values in y_train: 1144
Missing values in y_test: 660

Training target:
risk_level
Moderate Risk    2631
Low Risk          822
High Risk         809
Name: count, dtype: int64

Testing target:
risk_level
Moderate Risk    1487
Low Risk          562
High Risk         377
Name: count, dtype: int64


#### 1.1.3 Missing Values & Outliers Handling:

Missing values are checked separately for the input variables and the target variable.

Missing values in the target variable `risk_level` are removed because the classification model cannot learn from a record without a known risk category.

Missing values in the input features will be handled later through the preprocessing pipeline.

Outliers in numerical features such as yield, production, area, rainfall, and irrigation are investigated rather than automatically removed, because extreme values may represent genuine differences between districts, crops, seasons, or agricultural conditions.

In [199]:
# Check missing values in target
print("Missing values in risk_level:")
print(df["risk_level"].isnull().sum())

# Remove rows where target is missing
df = df.dropna(subset=["risk_level"]).copy()

print("\nShape after removing missing target:")
print(df.shape)

Missing values in risk_level:
0

Shape after removing missing target:
(8492, 53)


In [200]:
## check numerical outliers
numeric_columns = [
    "area",
    "production",
    "yield",
    "actual_rainfall",
    "normal_rainfall",
    "rainfall_deviation",
    "total_irrigated_area"
]

for col in numeric_columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_limit = Q1 - 1.5 * IQR
    upper_limit = Q3 + 1.5 * IQR

    outliers = ((df[col] < lower_limit) | (df[col] > upper_limit)).sum()

    print(f"{col}: {outliers} potential outliers")

area: 1613 potential outliers
production: 1679 potential outliers
yield: 940 potential outliers
actual_rainfall: 0 potential outliers
normal_rainfall: 0 potential outliers
rainfall_deviation: 0 potential outliers
total_irrigated_area: 0 potential outliers


### Outlier Decision

The IQR method identified potential outliers in **area (1613), production (1679), and yield (940)**.

These observations are retained because extreme values can naturally occur across different crops, districts, and seasons. No observations are removed based only on the IQR method. The impact of these extreme values will be considered during model evaluation.

### 1.1.4 Feature Engineering of X for Y

Feature engineering is performed to prepare the input variables for predicting
the agricultural risk level.

The available dataset contains agricultural, rainfall and irrigation information.
Additional historical features are generated from the existing data to capture
previous performance, trends, variability and changes over time.

#### 1.1.4.1 Feature Generation

Historical features are generated using the `year`, district, crop and season
information.

Features such as previous-year yield, historical average yield, yield trend,
yield variability, rainfall anomaly, rainfall variability, previous-year
irrigation, irrigation change and irrigation coverage are created to represent
agricultural stress and historical conditions.

#### 1.1.4.2 Feature Selection

The following features are selected based on their relevance to agricultural
risk identification and their availability in the prepared dataset.

The selected features include historical crop performance, rainfall conditions,
irrigation conditions and agricultural context such as district, crop, crop type
and season.

In [201]:
features = [
    "previous_year_yield",
    "historical_avg_yield",
    "yield_trend",
    "yield_variability",
    "previous_year_production",
    "historical_avg_production",
    "previous_year_area",
    "area_change",
    "normal_rainfall",
    "historical_rainfall",
    "rainfall_anomaly",
    "rainfall_deviation",
    "rainfall_variability",
    "previous_year_irrigation",
    "historical_avg_irrigation",
    "irrigation_change",
    "irrigation_coverage",
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]

In [202]:
X = df[features]
y = df["risk_level"]

#### 1.1.4.3 Feature Type Identification

The selected input features contain both **categorical and numerical variables**.

Categorical features such as `district_name`, `crop_name`, `crop_type`, and `season` will be converted into numerical form using **One-Hot Encoding**.

Numerical features such as historical yield, production, area, rainfall and irrigation features will be used as numerical inputs. Scaling will be applied for distance-based algorithms such as **KNN and SVM**.

In [203]:
categorical_features = [
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]

numerical_features = [
    "previous_year_yield",
    "historical_avg_yield",
    "yield_trend",
    "yield_variability",
    "previous_year_production",
    "historical_avg_production",
    "previous_year_area",
    "area_change",
    "normal_rainfall",
    "historical_rainfall",
    "rainfall_anomaly",
    "rainfall_deviation",
    "rainfall_variability",
    "previous_year_irrigation",
    "historical_avg_irrigation",
    "irrigation_change",
    "irrigation_coverage"
]

print("Categorical features:")
print(categorical_features)

print("\nNumerical features:")
print(numerical_features)

Categorical features:
['district_name', 'crop_name', 'crop_type', 'season']

Numerical features:
['previous_year_yield', 'historical_avg_yield', 'yield_trend', 'yield_variability', 'previous_year_production', 'historical_avg_production', 'previous_year_area', 'area_change', 'normal_rainfall', 'historical_rainfall', 'rainfall_anomaly', 'rainfall_deviation', 'rainfall_variability', 'previous_year_irrigation', 'historical_avg_irrigation', 'irrigation_change', 'irrigation_coverage']


#### 1.1.4.4 Missing Value Handling

Missing numerical values will be replaced using the **median** calculated from
the training data.

Missing categorical values will be replaced using the **most frequent value**
from the training data.

This approach prevents missing values from affecting model training while
ensuring that information from the test dataset does not influence the
preprocessing process.

#### 1.1.4.5 Encoding

Categorical features such as `district_name`, `crop_name`, `crop_type`, and
`season` are converted into numerical values using **One-Hot Encoding**.

One-hot encoding creates separate binary columns for each category, allowing
machine learning classification algorithms to process the categorical data.

In [204]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("encoder", OneHotEncoder(
            handle_unknown="ignore",
            sparse_output=False
        ))
    ]
)

numerical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="median"))
    ]
)

In [205]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numerical_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

print("Preprocessing pipeline created successfully.")

Preprocessing pipeline created successfully.


#### 1.1.4.6 Scaling
For numerical features

In [214]:
from sklearn.preprocessing import StandardScaler

numerical_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler())
    ]
)

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numerical_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

print("Preprocessing pipeline created successfully.")

print(preprocessor)

Preprocessing pipeline created successfully.
ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('scaler', StandardScaler())]),
                                 ['previous_year_yield', 'historical_avg_yield',
                                  'yield_trend', 'yield_variability',
                                  'previous_year_production',
                                  'historical_avg_production',
                                  'previous_year_area', 'area_change',
                                  'normal_rainfall', 'historical_rainfall',
                                  'rainfall_anomaly', 'rainfall_deviation',
                                  'rainfall_variability',
                                  'previous_year_irrigation',
                                  'historical_avg_irrigation',
      

### 1.2 Model + Define, Train & Study

For the Agricultural Risk Classification problem, classification algorithms are
selected based on the categorical target variable `risk_level`.

The following classification algorithms are considered:

- **Logistic Regression**
- **K-Nearest Neighbors (KNN)**
- **Naive Bayes**
- **Support Vector Machine (SVM)**
- **Decision Tree Classifier**
- **Random Forest Classifier**

XGBoost is not used in this project due to installation and resource constraints.

The selected models are loaded using the Python `sklearn` library and trained
using the training data. Their performance will later be evaluated using the
test data.

In [215]:
## Import Classification Models
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [216]:
## Define Models
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "KNN": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB(),
    "SVM": SVC(probability=True),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42)
}

print("Classification models defined successfully.")

Classification models defined successfully.


#### 1.2.4 Train Models

In [217]:
trained_models = {}

for name, model in models.items():

    pipeline = Pipeline(
        steps=[
            ("preprocessor", preprocessor),
            ("model", model)
        ]
    )

    pipeline.fit(X_train, y_train)

    trained_models[name] = pipeline

    print(f"{name} trained successfully.")

Logistic Regression trained successfully.
KNN trained successfully.
Naive Bayes trained successfully.
SVM trained successfully.
Decision Tree trained successfully.
Random Forest trained successfully.


### 1.3 Trained Model Predictions & Evaluations

The trained classification models are used to predict the `risk_level` for the
unseen test data.

The model performance is evaluated using classification metrics such as
**Accuracy, Precision, Recall and F1-Score** along with the **Confusion Matrix**.

The training and testing performance of each model is compared to identify
underfitting and overfitting using the **Bias-Variance Trade-off**.

Finally, the best classification model is selected based on its overall
performance on the test dataset, with greater importance given to Precision,
Recall and F1-Score along with Accuracy.

#### 1.3.1 Predictions on Train and Test Data

Predictions are generated using all trained classification models on both
training and test datasets.

The training predictions are used to understand model fitting, while
test predictions are used to evaluate how well the models perform on
unseen data.

In [ ]:
predictions = {}

for name, model in trained_models.items():

    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    predictions[name] = {
        "y_train_pred": y_train_pred,
        "y_test_pred": y_test_pred
    }

    print(f"{name} predictions completed.")

#### 1.3.2 Classification Evaluation Metrics

In [ ]:
from sklearn.metrics import accuracy_score, f1_score

results = []

for name in predictions:

    y_train_pred = predictions[name]["y_train_pred"]
    y_test_pred = predictions[name]["y_test_pred"]

    train_f1 = f1_score(
        y_train,
        y_train_pred,
        average="weighted",
        zero_division=0
    )

    test_f1 = f1_score(
        y_test,
        y_test_pred,
        average="weighted",
        zero_division=0
    )

    train_accuracy = accuracy_score(y_train, y_train_pred)
    test_accuracy = accuracy_score(y_test, y_test_pred)

    results.append({
        "Models": name,
        "TrainScore (F1)": train_f1,
        "TestScore (F1)": test_f1,
        "Train Accuracy": train_accuracy,
        "Test Accuracy": test_accuracy
    })

In [ ]:
results_df = pd.DataFrame(results)

results_df.round(3)

#### 1.3.3 Bias-Variance Trade-off

In [ ]:
def fit_type(train_score, test_score):

    difference = train_score - test_score

    if train_score < 0.70 and test_score < 0.70:
        return "Underfit"

    elif difference > 0.10:
        return "Overfit"

    else:
        return "Goodfit"


results_df["Bias-Variance (Fit)"] = results_df.apply(
    lambda row: fit_type(
        row["TrainScore (F1)"],
        row["TestScore (F1)"]
    ),
    axis=1
)

results_df.round(3)

#### 1.3.4 Final Classification Evaluation Table

In [ ]:
evaluation_table = pd.DataFrame(results_df)

evaluation_table[
    [
        "Models",
        "TrainScore (F1)",
        "TestScore (F1)",
        "Bias-Variance (Fit)"
    ]
]

In [ ]:
evaluation_table = evaluation_table.round({
    "TrainScore (F1)": 2,
    "TestScore (F1)": 2
})

evaluation_table

#### 1.3.5 Pick the best Model

In [ ]:
# Select the model with the highest Test F1-Score

best_model_name = evaluation_table.loc[
    evaluation_table["TestScore (F1)"].idxmax(),
    "Models"
]

print("Best Classification Model:", best_model_name)

In [ ]:
best_model_result = evaluation_table[
    evaluation_table["Models"] == best_model_name
]

best_model_result

Based on the evaluation results, the Decision Tree achieved a Test
F1-Score of 0.80. However, its Train F1-Score was 1.00, while the
Test F1-Score was 0.80, indicating an overfitting issue.

Therefore, the Decision Tree was not selected solely based on its
training performance. The final classification model is selected by
considering the Test F1-Score together with the Bias-Variance (Fit)
result. A model with a good balance between training and testing
performance is preferred for better generalization on unseen data.

#### 1.4 Hyperparameter Tuning

Hyperparameter tuning is performed to improve the performance of the
classification models and reduce overfitting.

The hyperparameters of the selected model are adjusted and evaluated
using the Test F1-Score. The tuned model is compared with the original
model to check whether its performance and generalization have improved.

If tuning does not provide a meaningful improvement, the original model
is retained as the final model.

In [ ]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    "model__max_depth": [3, 5, 7, 10, None],
    "model__min_samples_split": [2, 5, 10],
    "model__min_samples_leaf": [1, 2, 5, 10]
}

dt_model = trained_models["Decision Tree"]

grid_search = GridSearchCV(
    estimator=dt_model,
    param_grid=param_grid,
    cv=5,
    scoring="f1_weighted",
    n_jobs=-1
)

grid_search.fit(X_train, y_train)

print("Best Parameters:")
print(grid_search.best_params_)

print("\nBest Cross-Validation F1-Score:")
print(round(grid_search.best_score_, 3))

In [ ]:
tuned_dt = grid_search.best_estimator_

print("Tuned Decision Tree model created successfully.")

In [ ]:
from sklearn.metrics import f1_score, accuracy_score

y_train_pred_tuned = tuned_dt.predict(X_train)
y_test_pred_tuned = tuned_dt.predict(X_test)

train_f1_tuned = f1_score(
    y_train,
    y_train_pred_tuned,
    average="weighted",
    zero_division=0
)

test_f1_tuned = f1_score(
    y_test,
    y_test_pred_tuned,
    average="weighted",
    zero_division=0
)

train_accuracy_tuned = accuracy_score(y_train, y_train_pred_tuned)
test_accuracy_tuned = accuracy_score(y_test, y_test_pred_tuned)

print("Tuned Decision Tree")
print("--------------------------")
print("Train F1:", round(train_f1_tuned, 3))
print("Test F1:", round(test_f1_tuned, 3))
print("Train Accuracy:", round(train_accuracy_tuned, 3))
print("Test Accuracy:", round(test_accuracy_tuned, 3))

### 1.5 Saving Model & Real-Time Prediction

#### 1.5.1 Select Final Model

Based on the model evaluation results, Decision Tree Classifier was selected as the final classification model for agricultural risk prediction.

The model achieved the highest Test F1-Score of 0.80 among the evaluated classification models. The trained Decision Tree pipeline includes the preprocessing steps for numerical and categorical features along with the classification model.

In [ ]:
final_model = tuned_dt

print("Final model:", type(final_model.named_steps["model"]).__name__)

#### 1.5.2 Save the Best Model

In [ ]:
import joblib

best_model = tuned_dt

joblib.dump(best_model, "crop_risk_classification_model.pkl")

print("Best classification model saved successfully.")

#### 1.5.3 Load the Saved Model

In [ ]:
loaded_model = joblib.load("crop_risk_classification_model.pkl")

print("Saved classification model loaded successfully.")

#### 1.5.4 Real-Time Prediction

In [55]:
# ============================================================
# REAL-TIME AGRICULTURAL RISK PREDICTION
# ============================================================

import pandas as pd


# ------------------------------------------------------------
# 1. Load Dataset
# ------------------------------------------------------------

data_path = r"C:\Users\SVS\Desktop\ML Project\data\ml_ready\Final_dataset.csv"

df = pd.read_csv(data_path)


# ------------------------------------------------------------
# 2. Farmer Inputs
# ------------------------------------------------------------

print("Enter details for agricultural risk prediction")
print("------------------------------------------------")

district_name = input("Enter District: ").strip().lower()
crop_name = input("Enter Crop Name: ").strip().lower()
season = input("Enter Season: ").strip().lower()
area = float(input("Enter Area of Land: "))


# ------------------------------------------------------------
# 3. Prepare Text Columns
# ------------------------------------------------------------

df["district_name"] = (
    df["district_name"]
    .astype(str)
    .str.strip()
    .str.lower()
)

df["crop_name"] = (
    df["crop_name"]
    .astype(str)
    .str.strip()
    .str.lower()
)

df["crop_type"] = (
    df["crop_type"]
    .astype(str)
    .str.strip()
    .str.lower()
)

df["season"] = (
    df["season"]
    .astype(str)
    .str.strip()
    .str.lower()
)


# ------------------------------------------------------------
# 4. Sort Dataset
# ------------------------------------------------------------

df = df.sort_values(
    ["district_name", "crop_name", "season", "year"]
).copy()

group_cols = [
    "district_name",
    "crop_name",
    "season"
]


# ------------------------------------------------------------
# 5. Previous Year Features
# ------------------------------------------------------------

df["previous_year_yield"] = (
    df.groupby(group_cols)["yield"]
      .shift(1)
)

df["previous_year_production"] = (
    df.groupby(group_cols)["production"]
      .shift(1)
)

df["previous_year_area"] = (
    df.groupby(group_cols)["area"]
      .shift(1)
)

df["previous_year_irrigation"] = (
    df.groupby(group_cols)["total_irrigated_area"]
      .shift(1)
)


# ------------------------------------------------------------
# 6. Historical Average Features
# ------------------------------------------------------------

df["historical_avg_yield"] = (
    df.groupby(group_cols)["yield"]
      .transform(
          lambda x: x.shift(1).expanding().mean()
      )
)

df["historical_avg_production"] = (
    df.groupby(group_cols)["production"]
      .transform(
          lambda x: x.shift(1).expanding().mean()
      )
)

df["historical_avg_irrigation"] = (
    df.groupby(group_cols)["total_irrigated_area"]
      .transform(
          lambda x: x.shift(1).expanding().mean()
      )
)


# ------------------------------------------------------------
# 7. Yield Features
# ------------------------------------------------------------

df["yield_trend"] = (
    df["previous_year_yield"]
    - df["historical_avg_yield"]
)

df["yield_variability"] = (
    df.groupby(group_cols)["yield"]
      .transform(
          lambda x:
          x.shift(1)
           .rolling(3, min_periods=2)
           .std()
      )
)


# ------------------------------------------------------------
# 8. Area Feature
# ------------------------------------------------------------

df["area_change"] = (
    df["area"]
    - df["previous_year_area"]
)


# ------------------------------------------------------------
# 9. Rainfall Features
# ------------------------------------------------------------

df["historical_rainfall"] = (
    df.groupby(group_cols)["actual_rainfall"]
      .transform(
          lambda x:
          x.shift(1)
           .expanding()
           .mean()
      )
)

df["rainfall_anomaly"] = (
    df["actual_rainfall"]
    - df["normal_rainfall"]
)

df["rainfall_variability"] = (
    df.groupby(group_cols)["actual_rainfall"]
      .transform(
          lambda x:
          x.shift(1)
           .rolling(3, min_periods=2)
           .std()
      )
)


# ------------------------------------------------------------
# 10. Irrigation Features
# ------------------------------------------------------------

df["irrigation_change"] = (
    df["total_irrigated_area"]
    - df["previous_year_irrigation"]
)

df["irrigation_coverage"] = (
    df["total_irrigated_area"]
    / df["area"]
)


# ------------------------------------------------------------
# 11. Find Matching Records
# ------------------------------------------------------------

matching_data = df[
    (df["district_name"] == district_name) &
    (df["crop_name"] == crop_name) &
    (df["season"] == season)
]


# ------------------------------------------------------------
# 12. Check Matching Data
# ------------------------------------------------------------

if matching_data.empty:

    print("\nNo matching data found.")
    print("Please check District, Crop Name and Season.")


else:

    # --------------------------------------------------------
    # 13. Select Latest Available Record
    # --------------------------------------------------------

    matching_data = matching_data.sort_values("year")

    selected_data = matching_data.iloc[-1]


    # --------------------------------------------------------
    # 14. Create Model Input
    # --------------------------------------------------------

    sample_input = pd.DataFrame({

        "previous_year_yield": [
            selected_data["previous_year_yield"]
        ],

        "historical_avg_yield": [
            selected_data["historical_avg_yield"]
        ],

        "yield_trend": [
            selected_data["yield_trend"]
        ],

        "yield_variability": [
            selected_data["yield_variability"]
        ],

        "previous_year_production": [
            selected_data["previous_year_production"]
        ],

        "historical_avg_production": [
            selected_data["historical_avg_production"]
        ],

        "previous_year_area": [
            selected_data["previous_year_area"]
        ],

        "area_change": [
            selected_data["area_change"]
        ],

        "normal_rainfall": [
            selected_data["normal_rainfall"]
        ],

        "historical_rainfall": [
            selected_data["historical_rainfall"]
        ],

        "rainfall_anomaly": [
            selected_data["rainfall_anomaly"]
        ],

        "rainfall_deviation": [
            selected_data["rainfall_deviation"]
        ],

        "rainfall_variability": [
            selected_data["rainfall_variability"]
        ],

        "previous_year_irrigation": [
            selected_data["previous_year_irrigation"]
        ],

        "historical_avg_irrigation": [
            selected_data["historical_avg_irrigation"]
        ],

        "irrigation_change": [
            selected_data["irrigation_change"]
        ],

        "irrigation_coverage": [
            selected_data["irrigation_coverage"]
        ],

        "district_name": [
            district_name
        ],

        "crop_name": [
            crop_name
        ],

        "crop_type": [
            selected_data["crop_type"]
        ],

        "season": [
            season
        ]
    })


    # --------------------------------------------------------
    # 15. Predict Risk
    # --------------------------------------------------------

    predicted_risk = loaded_model.predict(
        sample_input
    )[0]


    # --------------------------------------------------------
    # 16. Risk Probability
    # --------------------------------------------------------

    risk_probabilities = loaded_model.predict_proba(
        sample_input
    )[0]

    risk_classes = loaded_model.classes_

    probability_dict = dict(
        zip(
            risk_classes,
            risk_probabilities
        )
    )

    predicted_probability = (
        probability_dict[predicted_risk] * 100
    )


    # --------------------------------------------------------
    # 17. Calculate Risk Factors
    # --------------------------------------------------------

    # Rainfall condition

    normal_rainfall = selected_data[
        "normal_rainfall"
    ]

    rainfall_deviation = selected_data[
        "rainfall_deviation"
    ]

    if normal_rainfall != 0:

        rainfall_score = (
            abs(rainfall_deviation)
            / abs(normal_rainfall)
        )

    else:

        rainfall_score = 0


    # Historical instability

    historical_avg_yield = selected_data[
        "historical_avg_yield"
    ]

    yield_variability = selected_data[
        "yield_variability"
    ]

    if historical_avg_yield != 0:

        instability_score = (
            abs(yield_variability)
            / abs(historical_avg_yield)
        )

    else:

        instability_score = 0


    # Irrigation condition

    historical_avg_irrigation = selected_data[
        "historical_avg_irrigation"
    ]

    irrigation_change = selected_data[
        "irrigation_change"
    ]

    if historical_avg_irrigation != 0:

        irrigation_score = (
            max(0, -irrigation_change)
            / abs(historical_avg_irrigation)
        )

    else:

        irrigation_score = 0


    # Historical yield trend

    yield_trend = selected_data[
        "yield_trend"
    ]

    if historical_avg_yield != 0:

        yield_trend_score = (
            max(0, -yield_trend)
            / abs(historical_avg_yield)
        )

    else:

        yield_trend_score = 0


    # --------------------------------------------------------
    # 18. Store Risk Factor Scores
    # --------------------------------------------------------

    factor_scores = {

        "Rainfall condition":
            rainfall_score,

        "Historical instability":
            instability_score,

        "Irrigation condition":
            irrigation_score,

        "Historical yield trend":
            yield_trend_score
    }


    # Sort factors from highest to lowest

    sorted_factors = sorted(
        factor_scores.items(),
        key=lambda x: x[1],
        reverse=True
    )


    # --------------------------------------------------------
    # 19. Contribution Level
    # --------------------------------------------------------

    def get_contribution(rank, score):

        if score <= 0:
            return "Low contribution"

        elif rank == 0:
            return "High contribution"

        elif rank == 1:
            return "Medium contribution"

        else:
            return "Low contribution"


    # --------------------------------------------------------
    # 20. Risk Symbol
    # --------------------------------------------------------

    if predicted_risk.lower() == "high risk":

        risk_symbol = "🔴"

    elif predicted_risk.lower() == "moderate risk":

        risk_symbol = "🟠"

    else:

        risk_symbol = "🟢"


    # --------------------------------------------------------
    # 21. Display Final Result
    # --------------------------------------------------------

    print("\n===============================================")
    print("        AGRICULTURAL RISK PREDICTION")
    print("===============================================")

    print(
        "District:",
        district_name.title()
    )

    print(
        "Crop:",
        crop_name.title()
    )

    print(
        "Season:",
        season.title()
    )

    print(
        "Area:",
        area
    )

    print()

    print(
        risk_symbol,
        predicted_risk.upper()
    )

    print(
        "Risk Probability:",
        round(predicted_probability, 2),
        "%"
    )


    # --------------------------------------------------------
    # 22. Display All Risk Probabilities
    # --------------------------------------------------------

    print("\nRisk Probabilities:")

    for risk_class, probability in probability_dict.items():

        print(
            risk_class,
            ":",
            round(probability * 100, 2),
            "%"
        )


    # --------------------------------------------------------
    # 23. Display Main Risk Factors
    # --------------------------------------------------------

    print("\nMain Risk Factors:")

    for rank, (factor, score) in enumerate(
        sorted_factors
    ):

        contribution = get_contribution(
            rank,
            score
        )

        if factor == "Rainfall condition":

            icon = "🌧"

        elif factor == "Historical instability":

            icon = "🌾"

        elif factor == "Irrigation condition":

            icon = "💧"

        else:

            icon = "📈"

        print(
            icon,
            factor,
            "→",
            contribution
        )


    print("===============================================")

Enter details for agricultural risk prediction
------------------------------------------------


Enter District:  karimnagar
Enter Crop Name:  rice
Enter Season:  kharif
Enter Area of Land:  4


NameError: name 'loaded_model' is not defined

In [ ]:
# Check all unique values in the dataset

print("Unique Districts:")
print(sorted(df["district_name"].dropna().unique()))


print("\nUnique Crop Names:")
print(sorted(df["crop_name"].dropna().unique()))

print("\nUnique Crop Types:")
print(sorted(df["crop_type"].dropna().unique()))

print("\nUnique Seasons:")
print(sorted(df["season"].dropna().unique()))

In [256]:
# Selected features

features = [
    "previous_year_yield",
    "historical_avg_yield",
    "yield_trend",
    "yield_variability",
    "previous_year_production",
    "historical_avg_production",
    "previous_year_area",
    "area_change",
    "normal_rainfall",
    "historical_rainfall",
    "rainfall_anomaly",
    "rainfall_deviation",
    "rainfall_variability",
    "previous_year_irrigation",
    "historical_avg_irrigation",
    "irrigation_change",
    "irrigation_coverage",
    "district_name",
    "crop_name",
    "crop_type",
    "season"
]

# Remove rows where target is missing
df = df.dropna(subset=["risk_level"]).copy()

# Reset index after removing rows
df = df.reset_index(drop=True)

# Create X and y again
X = df[features].copy()
y = df["risk_level"].copy()

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (6688, 21)
y shape: (6688,)


In [257]:
# Chronological split

years = sorted(df["year_start"].unique())

split_index = int(len(years) * 0.80)

train_years = years[:split_index]
test_years = years[split_index:]

print("Training years:")
print(train_years)

print("\nTesting years:")
print(test_years)

Training years:
[np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020)]

Testing years:
[np.int64(2021), np.int64(2022)]


In [258]:
# Create train and test masks

train_mask = df["year_start"].isin(train_years)
test_mask = df["year_start"].isin(test_years)

print("Training records:", train_mask.sum())
print("Testing records:", test_mask.sum())

Training records: 3904
Testing records: 2784


In [259]:
# Splitting X and y

X_train = X.loc[train_mask].copy()
X_test = X.loc[test_mask].copy()

y_train = y.loc[train_mask].copy()
y_test = y.loc[test_mask].copy()

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (3904, 21)
X_test shape: (2784, 21)
y_train shape: (3904,)
y_test shape: (2784,)


In [260]:
print("Missing values in y_train:", y_train.isna().sum())
print("Missing values in y_test:", y_test.isna().sum())

print("\nRisk levels in training:")
print(y_train.value_counts())

print("\nRisk levels in testing:")
print(y_test.value_counts())

Missing values in y_train: 0
Missing values in y_test: 0

Risk levels in training:
risk_level
Moderate Risk    2507
Low Risk          738
High Risk         659
Name: count, dtype: int64

Risk levels in testing:
risk_level
Moderate Risk    1611
Low Risk          646
High Risk         527
Name: count, dtype: int64
